Communicate refinement flags in p:s:T

peterrum opened the pull request:
adopted from https://github.com/dealii/dealii/blob/969cfe8cd8be564d2b916811c62aa7f68aaaf53a/source/dofs/dof_handler.cc#L1439-L1460

marcfehling commented:
With `parallel::shared::Triangulation`, the user is supposed to set refine and coarse flags on all processes accordingly. Each process has the complete information about all cells. There should be no need to communicate flags.

I would be in favor of a check in Debug mode that throws an exception if flags differ. But communicating flags is not intended in the (original) design of the class. This is the same reason why you need to use `GridRefinement` functions for this class rather than `parallel::distributed::GridRefinement`: the former works on all cells, the latter works on locally owned cells. (related: #14072)

I had to use the workaround you referenced as users are only allowed to set the fe indices on locally owned cells. This restriction does not apply for refine and coarsen flags.

I find it confusing that we have three parallel Triangulation classes each with different capabilities, and I know that we are somehow working towards bringing all of them under one hood. If we want to work towards this consistency, we should also restrict setting refine and coarsen flags only to locally owned cells for all classes. Then the communication of the flags as suggested with this PR would make sense. Otherwise I oppose to this change.

I let the original authors of the parallel shared class comment on this.

peterrum commented:
@marcfehling I don't understand why you are blocking my PRs.

> I find it confusing that we have three parallel Triangulation classes each with different capabilities, and I know that we are somehow working towards bringing all of them under one hood. 

This won't change in a short time and is not doable by one person, in particular if the person(s) involved also have a job, which are absolutely not related to this topic.

> If we want to work towards this consistency, we should also restrict setting refine and coarsen flags only to locally owned cells for all classes.

This is somewhat independent of this PR. As far as I can see it, it is possible to set arbitrary  refinement/coarsening flags in the case of p:d:T and p:f:T and this is also done during distributed refinement. One would need to create a new version of `set_refine_flag()`/`set_coarsen_flag()`, which is called internally and allows to be used on artificial cells and ghost cells. The old one would be used by the user. Feel free to work towards that; you will have my support.

All I want is to be able 1) to select at one central place a triangulation type, 2) easily switch it on the fly and 3) have usercode that works for each triangulation without annoying dynamic-casts in the user code. Unfortunately this is NOT the case at the moment. In most parts of the library it is the case and the functions accept `Triangulation` and internally is decided what happens. But everything-refinement related does not allow such programming, forcing a lot of boiler code in the user code. E.g.:
```
if(p:d:T)
  p:d::GridRefinement::refine_and_coarsen_fixed_number(distributed_error_vector)
  p:d::GridRefinement::SolutionTransfer(...)
elif(p:s:T with artificial cells)
  given a distributed_error_vector, communicate it
  GridRefinement::refine_and_coarsen_fixed_number(error_vector)
  communicate refinement flags (as proposed in this PR)
  SolutionTransfer(...)
elif(p:s:T)
  given a distributed_error_vector, communicate it -> for debugging
  GridRefinement::refine_and_coarsen_fixed_number(error_vector)
  communicate refinement flags -> for debugging
  SolutionTransfer(...)
else
  GridRefinement::refine_and_coarsen_fixed_number(error_vector)
  SolutionTransfer(...)
```

The goal has to be that this logic is hidden to the user. One can reach this only with small steps.

You also need to consider that `p:s:T` is the only is the only parallel triangulation, which supports AMR. So all users who indeed want to support AMR for 1D (p:s:T), 2D (p:d:T), and 3D (p:d:T) are forced to implement the logic above in their code.

kronbichler commented on source/distributed/shared_tria.cc:
```diff
@@ -351,6 +351,51 @@ namespace parallel
+            cell->clear_refine_flag();
+            cell->clear_coarsen_flag();
+
+            Assert(
+              (refinement_configurations[cell->active_cell_index() * 2 + 0] >
+                   0 ?
+                 1 :
+                 0) +
+                  refinement_configurations[cell->active_cell_index() * 2 +
+                                            1] <=
+                1,
+              ExcMessage(
+                "Refinement/coarsening flags of cells are not consistent in parallel!"));
```
I don't understand how this tests that the flags are the same everywhere, as the initial comment in this function is suggesting. Isn't all that you check that do not set refinement and coarsen flags simultaneously anywhere? That is probably sufficient, but not what the comment says. (By doing the `allreduce` in place, you lose track of what that process stored originally. You could retrieve the information again from what is stored in the flags, but I'm not sure that comparison is what you want. Either way, I think it would be good to have a way to be able to only store flags on locally owned cells.)

peterrum replied:
Indeed, it only assures that for the same cell not a refinement flag and a coarsening flag are set. The reasoning is that if only flags of locally owned cells are set, this is the right behavior. In the other case, it is a good heuristic under the assumption that if any process sets a flag it is taken (if both refinement and coarsening flag are set, the situation is not clear).

drwells commented:
This reminds me of #12557 (with artificial cells): we have a bunch of inconsistencies which I find surprising between `p:s:T` and `p:d:T`.

The documentation states
```cpp
    /* [...]
     * As a consequence of storing the entire mesh on each processor, active
     * cells need to be flagged for refinement or coarsening consistently on
     * all processors if you want to adapt them, regardless of being classified
     * as locally owned, ghost or artificial.
     */
```
so if we proceed that will need to be updated.

The status quo is weird in that someone (AFAICT: I didn't try it) would need to set refinement flags on artificial cells to do AMR with `p:s::T` right now. I am in favor of changing the behavior of this class so that "`p:s:T` with artificial cells and exactly one level" works exactly the same way as "`p:d:T` with one exactly level".

How about we vote on that statement here: :+1: if you agree (so we do this PR and followup adjustments) and :-1: if you do not. That will resolve the impasse.

drwells commented:
@dealii/dealii if anyone else has opinions on `p:s:T` then please vote so we can either go ahead with this change or proceed in a different way.

drwells commented:
It looks like the consensus is to change the behavior of the class. Could we add a test to make sure the old version of setting flags still works?

peterrum commented:
> It looks like the consensus is to change the behavior of the class. Could we add a test to make sure the old version of setting flags still works?

There are tests for that already.

I have added a new test to test the new behavior. Furthermore, I have updated the documentation.
